[IPv6] Support flow exporter

lzhecheng opened the pull request:
Support IPv4 or IPv6 flow exporter address.

lxiaopei commented:
/test-e2e /test-conformance /test-networkpolicy

srikartati commented on build/yamls/antrea-aks.yml:
```diff
@@ -1199,9 +1199,10 @@ data:
-    # the flow exporter that sends IPFIX flow records of conntrack flows on OVS bridge. If no L4 transport proto is given,
-    # we consider tcp as default.
+    # Provide flow collector address as string with format <IP>:<port>[:<proto>], where proto is tcp or udp. IP can be IPv4
+    # or IPv6 and IPv6 address should be wrapped with [].
```
IP can be either IPv4 or IPv6; however, IPv6 address should be wrapped with [].

lzhecheng replied:
Updated.

srikartati commented on pkg/agent/flowexporter/exporter/exporter.go:
```diff
@@ -43,6 +43,20 @@ var (
 		"packetDeltaCount",
 		"octetDeltaCount",
 	}
+	IANAInfoElementsIPv6 = []string{
```
Don't we need AntreaInfoElementsIPv6 for destinationClusterIPv6, when clusterIP/svcCIDR is v6? Pod-To-Service flow with IPv6 will not have the clusterIP. 

As a side note, we probably need to improve unit test TestFlowExporter_sendDataRecord by considering values in the input flow to cover scenarios with IPv4 and IPv6 to catch these errors. Should be considered for future.

lzhecheng replied:
Thanks for reminding me. Updated.
Updated tests as well.

srikartati reviewed (commented):
My main comment is about destinationClusterIPv6. Otherwise, it looks good to me.

srikartati reviewed (commented):
@lzhecheng Currently we consider only the connections in [Antrea connection zone of conntrack table](https://github.com/vmware-tanzu/antrea/blob/78477c9fe3eccf855cac1fb0f58b82a7e3f8935f/pkg/agent/flowexporter/connections/connections.go#L237) and ignore connections in the default zone. 

There is a separate connection zone for IPv6. Don't we need to consider that connection zone if the stack is IPv6? Wondering if the flow exporter e2e test is passing with IPv6 stack or not.

lzhecheng commented:
> @lzhecheng Currently we consider only the connections in [Antrea connection zone of conntrack table](https://github.com/vmware-tanzu/antrea/blob/78477c9fe3eccf855cac1fb0f58b82a7e3f8935f/pkg/agent/flowexporter/connections/connections.go#L237) and ignore connections in the default zone.
> 
> There is a separate connection zone for IPv6. Don't we need to consider that connection zone if the stack is IPv6? Wondering if the flow exporter e2e test is passing with IPv6 stack or not.

I think I have it?
https://github.com/vmware-tanzu/antrea/pull/1541/files#diff-
fa3c4bf704e90acceacb369eed2fe23b739bd365e8d41ff087becab848371924R244

lzhecheng commented:
/test-windows-conformance
/test-windows-networkpolicy

srikartati reviewed (dismissed):
Thanks for making the changes. 
LGTM.

srikartati commented:
/test-ipv6-all
/test-ipv6-only-all

antoninbas commented on cmd/antrea-agent/agent.go:
```diff
@@ -298,9 +299,20 @@ func run(o *Options) error {
 
 	// Initialize flow exporter to start go routines to poll conntrack flows and export IPFIX flow records
 	if features.DefaultFeatureGate.Enabled(features.FlowExporter) {
+		match, err := regexp.MatchString("\\[.*\\]:.*", o.config.FlowCollectorAddr)
+		if err != nil {
+			return fmt.Errorf("Failed to parse FlowCollectorAddr: %s", o.config.FlowCollectorAddr)
+		}
+		svcCIDR := serviceCIDRNet
+		addrFamily := "ipv4"
+		if match {
+			svcCIDR = serviceCIDRNetv6
+			addrFamily = "ipv6"
+		}
```
is the following scenario not valid?
* the cluster is dual-stack, and Nodes have both an IPv4 address and an IPv6 address
* the primary address family for the cluster (serviceCIDR) is IPv6
* the collector runs outside the cluster (public IP) and only has an IPv4 address

Basically I'm asking whether it makes sense to use `o.config.FlowCollectorAddr` to determine the service address family? Why can't we just check whether `serviceCIDRNetv6 != ""`?

lzhecheng replied:
Thank you for the question. I think if user want to use flow exporter, `FlowCollectorAddr` should be correctly configured. Also, with AntreaProxy enabled, `serviceCIDRNetv6` is not needed.

lzhecheng replied:
`serviceCIDRNetv6` is not needed with AntreaProxy enabled while `FlowCollectorAddr` should be correctly configured by users.

antoninbas replied:
That only answers part of my question though.

What if my flow collector is running out-of-cluster and only has a public IPv4 address, but my cluster is dual-stack?

lzhecheng replied:
Sorry, I only paid attention to the sentences after "Basically"...
You are right. flowCollectorAddr is unreliable. Instead, I use nodeAddr now for judgement. Then in that situation, address family is IPv6 and FlowCollectorAddr is IPv4.

srikartati replied:
Does this mean we support pod-to-externalTraffic where pod is of IPv6 and external IP is IPv4? Will the IPv6 address be SNATed to IPv4 address?

lzhecheng replied:
Maybe not now. I created an ipfix-collector in an IPv6-only cluster. And I hard-coded its address into e2e test code in a dual-stack cluster, with IPv4 service CIDR. I ran the flow exporter test and checked logs of ipfix-collector. There's no IPv4 record.

srikartati replied:
In the above test, could antrea-agent pod (flow exporter) with IPv6 make a connection to the flow collector with IPv4 address? Wondering if that communication is possible at all?

lzhecheng replied:
I guess you mean antrea-agent pod with IPv4 as PodIP to connect flow collector pod with IPv6? The traffic is ok.

srikartati replied:
Please ignore my comment. Understood the scenario in Antonin's comment after offline discussion.

lzhecheng replied:
@srikartati Do you think this PR can be merged? The latest code uses `NodeAddr` to decide if in IPv4 or IPv6.

lzhecheng commented:
/test-ipv6-e2e
/test-windows-networkpolicy

lzhecheng commented:
/test-ipv6-e2e

lzhecheng commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all

antoninbas reviewed (commented):
Maybe I'm misunderstanding the scope of this PR but it seems it doesn't not support dual-stack. Maybe this is intentional, but in this case it would be best to indicate it in the commit message & PR description.

lzhecheng commented:
> Maybe I'm misunderstanding the scope of this PR but it seems it doesn't not support dual-stack. Maybe this is intentional, but in this case it would be best to indicate it in the commit message & PR description.

Updated commit message and PR description.

srikartati commented:
>https://github.com/vmware-tanzu/antrea/pull/1541#discussion_r526533169

@lzhecheng In the case of dual-stack, we need to send multiple templates one for IPv4 records and the other for IPv6 records. When sending IPv4 or IPv6 data records, the right template should be picked. I think this is what @antoninbas was suggesting.
I am ok with either: doing it in this PR or a different PR. If different PR, please raise an issue, so that we could finish it before next release. This helps in updating the [user doc](https://github.com/vmware-tanzu/antrea/blob/master/docs/network-flow-visibility.md) with IPv6 support. 

Regarding `nodeAddr`: Considering the issue with dual-stack explained in the [comment](https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/controller/noderoute/node_route_controller.go#L601), should we use `PodIPv4CIDR` and `PodIPv6CIDR` in `nodeConfig` struct  to determine the address family? if the PR doesn't handle the case of dual-stack, we can throw a warning that FlowExporter is not supported for dual-stack in the code itself and return.

lzhecheng commented:
> > [#1541 (comment)](https://github.com/vmware-tanzu/antrea/pull/1541#discussion_r526533169)
> 
> @lzhecheng In the case of dual-stack, we need to send multiple templates one for IPv4 records and the other for IPv6 records. When sending IPv4 or IPv6 data records, the right template should be picked. I think this is what @antoninbas was suggesting.
> I am ok with either: doing it in this PR or a different PR. If different PR, please raise an issue, so that we could finish it before the next release. This helps in updating the [user doc](https://github.com/vmware-tanzu/antrea/blob/master/docs/network-flow-visibility.md) with IPv6 support.
> 
> Regarding `nodeAddr`: Considering the issue with dual-stack explained in the [comment](https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/controller/noderoute/node_route_controller.go#L601), should we use `PodIPv4CIDR` and `PodIPv6CIDR` in `nodeConfig` struct to determine the address family? If this PR doesn't handle the case of dual-stack, we can throw a warning that FlowExporter is not supported for dual-stack in the code itself and return.

Thank you @srikartati ! I think we can finish dual-stack in this PR. You are right. `nodeAddr` is for current implementation and for dual-stack, pod CIDR is better.

antoninbas commented:
> > > [#1541 (comment)](https://github.com/vmware-tanzu/antrea/pull/1541#discussion_r526533169)
> > 
> > 
> > @lzhecheng In the case of dual-stack, we need to send multiple templates one for IPv4 records and the other for IPv6 records. When sending IPv4 or IPv6 data records, the right template should be picked. I think this is what @antoninbas was suggesting.
> > I am ok with either: doing it in this PR or a different PR. If different PR, please raise an issue, so that we could finish it before the next release. This helps in updating the [user doc](https://github.com/vmware-tanzu/antrea/blob/master/docs/network-flow-visibility.md) with IPv6 support.
> > Regarding `nodeAddr`: Considering the issue with dual-stack explained in the [comment](https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/controller/noderoute/node_route_controller.go#L601), should we use `PodIPv4CIDR` and `PodIPv6CIDR` in `nodeConfig` struct to determine the address family? If this PR doesn't handle the case of dual-stack, we can throw a warning that FlowExporter is not supported for dual-stack in the code itself and return.
> 
> Thank you @srikartati ! I think we can finish dual-stack in this PR. You are right. `nodeAddr` is for current implementation and for dual-stack, pod CIDR is better.

Note that using PodCIDR doesn't work for policyOnlyMode. You can use the new functions I'm introducing in https://github.com/vmware-tanzu/antrea/pull/1585: `config.IsIPv4Enabled` and `config.IsIPv6Enabled`

srikartati commented:
> 
> Note that using PodCIDR doesn't work for policyOnlyMode. You can use the new functions I'm introducing in #1585: `config.IsIPv4Enabled` and `config.IsIPv6Enabled`

util functions sound good. 
Not related to this PR: Wondering if Flow Exporter feature is applicable to `policyOnlyMode` setup or not. I believe we can skip it as we do not use Antrea dataplane. @antoninbas what do you think?

antoninbas commented:
> > Note that using PodCIDR doesn't work for policyOnlyMode. You can use the new functions I'm introducing in #1585: `config.IsIPv4Enabled` and `config.IsIPv6Enabled`
> 
> util functions sound good.
> Not related to this PR: Wondering if Flow Exporter feature is applicable to `policyOnlyMode` setup or not. I believe we can skip it as we do not use Antrea dataplane. @antoninbas what do you think?

We still commit all the connections for NP enforcement + AntreaProxy does the ClusterIP load-balancing, so I think we can export the info. I believe we want to be able to support FlowExporter for cloud-managed services. Is this not technically possible>

srikartati commented:
> > > Note that using PodCIDR doesn't work for policyOnlyMode. You can use the new functions I'm introducing in #1585: `config.IsIPv4Enabled` and `config.IsIPv6Enabled`
> > 
> > 
> > util functions sound good.
> > Not related to this PR: Wondering if Flow Exporter feature is applicable to `policyOnlyMode` setup or not. I believe we can skip it as we do not use Antrea dataplane. @antoninbas what do you think?
> 
> We still commit all the connections for NP enforcement + AntreaProxy does the ClusterIP load-balancing, so I think we can export the info. I believe we want to be able to support FlowExporter for cloud-managed services. Is this not technically possible?

Double checked policyOnly mode design doc. You are right. We can still get connections from the conntrack module.

lzhecheng commented:
> > > > [#1541 (comment)](https://github.com/vmware-tanzu/antrea/pull/1541#discussion_r526533169)
> > > 
> > > 
> > > @lzhecheng In the case of dual-stack, we need to send multiple templates one for IPv4 records and the other for IPv6 records. When sending IPv4 or IPv6 data records, the right template should be picked. I think this is what @antoninbas was suggesting.
> > > I am ok with either: doing it in this PR or a different PR. If different PR, please raise an issue, so that we could finish it before the next release. This helps in updating the [user doc](https://github.com/vmware-tanzu/antrea/blob/master/docs/network-flow-visibility.md) with IPv6 support.
> > > Regarding `nodeAddr`: Considering the issue with dual-stack explained in the [comment](https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/controller/noderoute/node_route_controller.go#L601), should we use `PodIPv4CIDR` and `PodIPv6CIDR` in `nodeConfig` struct to determine the address family? If this PR doesn't handle the case of dual-stack, we can throw a warning that FlowExporter is not supported for dual-stack in the code itself and return.
> > 
> > 
> > Thank you @srikartati ! I think we can finish dual-stack in this PR. You are right. `nodeAddr` is for current implementation and for dual-stack, pod CIDR is better.
> 
> Note that using PodCIDR doesn't work for policyOnlyMode. You can use the new functions I'm introducing in #1585: `config.IsIPv4Enabled` and `config.IsIPv6Enabled`

Thank you! They are very helpful.

lzhecheng commented:
/test-e2e
/test-ipv6-e2e
/test-ipv6-only-e2e

lzhecheng commented:
/test-e2e
/test-ipv6-e2e
/test-ipv6-only-e2e

lzhecheng commented:
/test-ipv6-e2e
/test-ipv6-only-e2e

lzhecheng commented:
/test-e2e

lzhecheng commented:
/test-e2e

lzhecheng commented:
/test-e2e
/test-ipv6-e2e
/test-ipv6-only-e2e

lzhecheng commented:
@srikartati I refactored the code. In dual-stack setup, antrea-agent does `initFlowExporter()` twice for IPv4 and IPv6. Considering possible conflict in `templateID`, IPv4 `templateID` is limited to [0, half uint16 max] and IPv6 one is (half uint16 max, uint16 max]. PTAL.

srikartati commented:
> @srikartati I refactored the code. In dual-stack setup, antrea-agent does `initFlowExporter()` twice for IPv4 and IPv6. Considering possible conflict in `templateID`, IPv4 `templateID` is limited to [0, half uint16 max] and IPv6 one is (half uint16 max, uint16 max]. PTAL.

Hi @lzhecheng , Thanks for working on this. 
You could use expProcess.NewTemplateID() when creating new template. In the case of dual stack, we just have to call that two times to have separate templates.

I have a question. Why do we need two separate connectionStore and exporter objects for the case of dual-stack?

lzhecheng commented:
> > @srikartati I refactored the code. In dual-stack setup, antrea-agent does `initFlowExporter()` twice for IPv4 and IPv6. Considering possible conflict in `templateID`, IPv4 `templateID` is limited to [0, half uint16 max] and IPv6 one is (half uint16 max, uint16 max]. PTAL.
> 
> Hi @lzhecheng , Thanks for working on this.
> You could use expProcess.NewTemplateID() when creating new template. In the case of dual stack, we just have to call that two times to have separate templates.
> 
> I have a question. Why do we need two separate connectionStore and exporter objects for the case of dual-stack?

Hello. Reason for seperate `connectionStore` and `exporter` is that when working on single-stack IPv6, I added a field `addrFamily` in `connectionStore` and `exporter` and in my mind, it can only be IPv4 or IPv6.

But actually maybe I can extend `addrFamily` to include `dual-stack`. I will give it a shot.

lzhecheng commented:
/test-e2e
/test-ipv6-e2e
/test-ipv6-only-e2e

lzhecheng commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all

lzhecheng commented:
/test-ipv6-e2e

lzhecheng commented:
/test-ipv6-e2e
/test-networkpolicy

lzhecheng commented:
/test-ipv6-e2e

antoninbas commented on pkg/agent/flowexporter/exporter/exporter.go:
```diff
@@ -63,16 +77,34 @@ var (
 		"egressNetworkPolicyName",
 		"egressNetworkPolicyNamespace",
 	}
+	AntreaInfoElementsIPv6 = []string{
```
nit: maybe we could define `antreaInfoElements` and use it in the definition of `AntreaInfoElementsIPv4` and `AntreaInfoElementsIPv6` to avoid repetition

lzhecheng replied:
Do you mean it is better to define 3 lists: `AntreaInfoElements` (common IE), `AntreaInfoElementsIPv4` (IPv4 specific IE) and `AntreaInfoElementsIPv6` (IPv6 specific IE)? Then append them in code below?

antoninbas replied:
yes that's what I mean. And the common IE does not need to be exported. https://play.golang.org/p/qDaN_AJzyOP

But if the order of the elements in the array matters, then feel free to ignore this comment

antoninbas commented on pkg/agent/flowexporter/flowrecords/flow_records.go:
```diff
@@ -71,12 +71,23 @@ func (fr *FlowRecords) ValidateAndUpdateStats(connKey flowexporter.ConnectionKey
-		if err != nil {
-			klog.Errorf("Error when executing callback for flow record")
-			return err
+func (fr *FlowRecords) ForAllFlowRecordsDo(callback flowexporter.FlowRecordCallBack, v4 bool, v6 bool, v4ID uint16, v6ID uint16) error {
```
maybe we should use `v4Enabled` and `v6Enabled` consistently?

lzhecheng replied:
Good idea.

antoninbas commented on test/e2e/flowexporter_test.go:
```diff
@@ -235,7 +236,11 @@ func checkRecordsWithPodIPs(t *testing.T, data *TestData, podAIP string, podBIP
+	if len(clusterInfo.podV4NetworkCIDR) != 0 && len(clusterInfo.podV6NetworkCIDR) != 0 {
+		expectedNumTemplateRecords = clusterInfo.numNodes * 2
+	}
+	assert.Equal(t, expectedNumTemplateRecords, templateRecords, "Each agent should send out template record")
```
```suggestion
	assert.Equal(t, expectedNumTemplateRecords, templateRecords, "Each agent should send out a template record per supported family address")

```

antoninbas commented on test/e2e/flowexporter_test.go:
```diff
@@ -31,10 +31,8 @@ import (
-	skipIfIPv6Cluster(t)
-	skipIfNotIPv4Cluster(t)
-	data, err := setupTestWithIPFIXCollector(t)
+	// Should I add skipBenchmark as this runs iperf?
+	data, err, isIPv6 := setupTestWithIPFIXCollector(t)
```
so we now have test coverage for IPv4, IPv6 and dual-stack?

lzhecheng replied:
Mostly. Except that I don't have a good idea how to test the case you mentioned:
An `ipfix-collector` is outside of a dual-stack cluster and its IP family is different from the main IP family (same as node's internal-ip) of the cluster. Pods generated in e2e should be in a cluster and `ipfix-collector` Pod uses host network so it only has one IP.

antoninbas replied:
@lzhecheng no worries about this. Right now we always deploy the ELK stack inside the cluster and the aggregator will of course be deployed in the cluster. I just wanted to make sure that we didn't make any assumption as to the address family of the collector and looks like it is the case now.

antoninbas commented on test/integration/agent/flowexporter_test.go:
```diff
@@ -159,9 +159,10 @@ func TestConnectionStoreAndFlowRecords(t *testing.T) {
+	connsLens, err := connStore.Poll()
 	require.Nil(t, err, fmt.Sprintf("Failed to add connections to connection store: %v", err))
-	assert.Equal(t, connsLen, len(testConns), "expected connections should be equal to number of testConns")
+	assert.Equal(t, len(connsLens), 1, "length of connsLens is expected to be 1")
+	assert.Equal(t, connsLens[0], len(testConns), "expected connections should be equal to number of testConns")
```
there is an `assert.Len` available

lzhecheng replied:
Updated.

antoninbas reviewed (commented):
overall LGTM now, seems that we support all possible cases correctly {v4, v6, dual-stack} X {v4 collector, v6 collector}

srikartati commented on pkg/agent/flowexporter/flowrecords/flow_records.go:
```diff
@@ -71,12 +71,23 @@ func (fr *FlowRecords) ValidateAndUpdateStats(connKey flowexporter.ConnectionKey
-			return err
+func (fr *FlowRecords) ForAllFlowRecordsDo(callback flowexporter.FlowRecordCallBack, v4 bool, v6 bool, v4ID uint16, v6ID uint16) error {
+	if v4 {
+		for k, v := range fr.recordsMap {
```
could we add isv4 field to the `FlowRecord` struct? This can be populated through the call back function `addOrUpdateFlowRecord` when creating new record. This may save sending multiple args in flow_record.go functions and exporter.go functions including template IDs.

lzhecheng replied:
Good idea. Updated. PTAL.

srikartati commented on pkg/agent/flowexporter/exporter/exporter.go:
```diff
@@ -163,44 +199,61 @@ func (exp *flowExporter) initFlowExporter(collector net.Addr) error {
+		if err != nil {
+			return err
+		}
+		klog.V(2).Infof("Initialized flow exporter IPv4 and sent %d bytes size of template record", sentBytes)
```
```suggestion
		klog.V(2).Infof("Initialized flow exporter for IPv4 flow records and sent %d bytes size of template record", sentBytes)
```

srikartati commented on pkg/agent/flowexporter/exporter/exporter.go:
```diff
@@ -163,44 +199,61 @@ func (exp *flowExporter) initFlowExporter(collector net.Addr) error {
+		if err != nil {
+			return err
+		}
+		klog.V(2).Infof("Initialized flow exporter IPv6 and sent %d bytes size of template record", sentBytes)
```
same here.

lzhecheng replied:
Updated.

srikartati commented on pkg/agent/flowexporter/connections/connections.go:
```diff
@@ -234,26 +240,36 @@ func (cs *ConnectionStore) Poll() (int, error) {
+			cs.addOrUpdateConn(conn)
+		}
+		connsLens = append(connsLens, len(filteredConnsList))
+		filteredConnsList = nil
```
Not your PR. This can be removed. Thanks.

lzhecheng replied:
Updated.

srikartati commented on pkg/agent/flowexporter/connections/connections.go:
```diff
@@ -234,26 +240,36 @@ func (cs *ConnectionStore) Poll() (int, error) {
+		}
+		connsLens = append(connsLens, len(filteredConnsList))
+		filteredConnsList = nil
+		metrics.TotalConnectionsInConnTrackTable.Set(float64(totalConns))
+		maxConns, err := cs.connDumper.GetMaxConnections()
+		if err != nil {
+			return []int{}, err
+		}
+		metrics.MaxConnectionsInConnTrackTable.Set(float64(maxConns))
```
Nit: this can be done outside the for loop with zones.

lzhecheng replied:
Updated.

srikartati reviewed (commented):
Thanks for working on this. The overall approach looks good to me; do this by polling conntrack module for multiple zones separately for dual stack makes sense to me.

lzhecheng commented:
/test-e2e
/test-ipv6-e2e
/test-ipv6-only-e2e

lzhecheng commented on pkg/agent/flowexporter/connections/connections.go:
```diff
@@ -221,7 +227,7 @@ func (cs *ConnectionStore) ForAllConnectionsDo(callback flowexporter.ConnectionM
 // Poll calls into conntrackDumper interface to dump conntrack flows
 // TODO: As optimization, only poll invalid/closed connections during every poll, and poll the established connections right before the export.
-func (cs *ConnectionStore) Poll() (int, error) {
+func (cs *ConnectionStore) Poll() ([]int, error) {
```
Make sense. Updated.

lzhecheng commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all

lzhecheng commented:
/test-conformance 
/test-ipv6-e2e
/test-ipv6-networkpolicy

srikartati commented on pkg/agent/flowexporter/exporter/exporter.go:
```diff
@@ -234,16 +272,24 @@ func (exp *flowExporter) sendTemplateSet(templateSet ipfix.IPFIXSet, templateID
 }
 
-func (exp *flowExporter) sendDataSet(dataSet ipfix.IPFIXSet, record flowexporter.FlowRecord, templateID uint16) error {
+func (exp *flowExporter) sendDataSet(dataSet ipfix.IPFIXSet, record flowexporter.FlowRecord, templateID uint16, isIPv6 bool) error {
```
Do we need the args `templateID` and `isIPv6`? I think`record flowexporter.FlowRecord` and `exp *flowExporter` should have these already.

lzhecheng replied:
Yes. Removed.

srikartati commented on pkg/agent/flowexporter/exporter/exporter.go:
```diff
@@ -195,9 +227,15 @@ func (exp *flowExporter) sendFlowRecords() error {
 }
 
-func (exp *flowExporter) sendTemplateSet(templateSet ipfix.IPFIXSet, templateID uint16) (int, error) {
+func (exp *flowExporter) sendTemplateSet(templateSet ipfix.IPFIXSet, templateID uint16, isIPv6 bool) (int, error) {
```
templateID is not needed as we assign the templateID to `exp *flowExporter` in L179 and L189 and can use that inside the function based on `isIPv6`. Is my understanding correct?

lzhecheng replied:
You are right. Updated.

srikartati commented on pkg/agent/flowexporter/exporter/exporter.go:
```diff
@@ -334,6 +384,13 @@ func (exp *flowExporter) sendDataSet(dataSet ipfix.IPFIXSet, record flowexporter
+				ie.Value = record.Conn.TupleOrig.DestinationAddress
+			} else {
+				// Same as destinationClusterIPv4.
+				ie.Value = net.IP{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
```
nit: simplified version net.IP.Parse("::0")

lzhecheng replied:
It seems `"::"` works as well. Updated.

srikartati commented on pkg/agent/flowexporter/types.go:
```diff
@@ -70,4 +70,6 @@ type FlowRecord struct {
 	PrevBytes          uint64
 	PrevReversePackets uint64
 	PrevReverseBytes   uint64
+	V4Enabled          bool
```
Just need one field isIPv6 as one flow record corresponds to either v4 or v6.

lzhecheng replied:
Updated.

srikartati commented on pkg/agent/flowexporter/flowrecords/flow_records.go:
```diff
@@ -35,9 +35,32 @@ func NewFlowRecords(connStore *connections.ConnectionStore) *FlowRecords {
 
 // BuildFlowRecords builds the flow record map from connection map in connection store
-func (fr *FlowRecords) BuildFlowRecords() error {
+func (fr *FlowRecords) BuildFlowRecords(v4Enabled bool, v6Enabled bool) error {
```
We do not need the args. From IP in key, we can say if it is v4 or v6.

lzhecheng replied:
Updated.

lzhecheng commented:
/test-e2e
/test-ipv6-e2e
/test-ipv6-only-e2e

lzhecheng commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all

srikartati commented on pkg/agent/flowexporter/flowrecords/flow_records.go:
```diff
@@ -36,8 +38,35 @@ func NewFlowRecords(connStore *connections.ConnectionStore) *FlowRecords {
+		return nil
+	}
+
 	// fr.addOrUpdateFlowRecord method does not return any error, hence no error handling required.
```
Change the comment that still says the method name of flow record rather than function name.

lzhecheng replied:
Indeed. `fr.addOrUpdateFlowRecord` -> `addOrUpdateFlowRecord`

srikartati reviewed (dismissed):
Thanks for addressing comments. 
LGTM.

lzhecheng commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all

lzhecheng commented:
/test-conformance

lzhecheng commented:
/test-e2e

lzhecheng commented:
/test-ipv6-only-conformance

lzhecheng commented:
/test-ipv6-only-conformance

antoninbas commented on pkg/agent/flowexporter/connections/connections.go:
```diff
@@ -219,9 +225,10 @@ func (cs *ConnectionStore) ForAllConnectionsDo(callback flowexporter.ConnectionM
 }
 
-// Poll calls into conntrackDumper interface to dump conntrack flows
+// Poll calls into conntrackDumper interface to dump conntrack flows. In a dual-stack setup, there will be two connsLens
+// get returned.
```
suggestion: Poll calls into conntrackDumper interface to dump conntrack flows. It returns the number of connections  for each address family, as a slice. In dual-stack clusters, the slice will contain 2 values (number of IPv4 connections first, then number of IPv6 connections).

lzhecheng replied:
Updated.

antoninbas commented on pkg/agent/flowexporter/exporter/exporter.go:
```diff
@@ -44,36 +42,44 @@ var (
 		"reverseOctetDeltaCount",
 	}
-	AntreaInfoElements = []string{
+	AntreaInfoElementsCommon = []string{
```
does not need to be exported I think... could be `antreaInfoElementsCommon`

lzhecheng replied:
Indeed. Updated.

antoninbas commented on pkg/agent/flowexporter/exporter/exporter_test.go:
```diff
@@ -21,147 +21,213 @@ import (
+
+func getElemList(ianaIE []string, antreaIE []string) []*ipfixentities.InfoElementWithValue {
+	// Following consists of all elements that are in IANAInfoElements and AntreaInfoElements (globals)
+	// Need only element name and other are dummys
```
s/other are dummys/other fields are set to dummy values

lzhecheng replied:
Updated.

antoninbas reviewed (commented):
couple nits, otherwise LGTM

lzhecheng commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all

lzhecheng commented:
/test-conformance
/test-networkpolicy

lzhecheng commented:
Thank you a lot for the effort to review! @antoninbas @srikartati 
Merging this PR.
